libpod/runtime_img.go

```diff
@@ -123,41 +211,107 @@ func (r *Runtime) PullImage(imgName string, allTags bool, signaturePolicyPath st
 		}
-		// to pull all the images stored in one tar file
-		for i := range manifest {
-			if manifest[i].RepoTags != nil {
-				images = append(images, manifest[i].RepoTags[0])
-			} else {
-				// create an image object and use the hex value of the digest as the image ID
-				// for parsing the store reference
-				newImg, err := srcRef.NewImage(sc)
+		// to pull the first image stored in the tar file
+		if len(manifest) == 0 {
+			// create an image object and use the hex value of the digest as the image ID
+			// for parsing the store reference
+			newImg, err := srcRef.NewImage(sc)
+			if err != nil {
+				return nil, err
+			}
+			defer newImg.Close()
+			digest := newImg.ConfigInfo().Digest
+			if err := digest.Validate(); err == nil {
+				pullInfo, err := r.getPullStruct(srcRef, "@"+digest.Hex())
 				if err != nil {
-					return err
-				}
-				defer newImg.Close()
-				digest := newImg.ConfigInfo().Digest
-				if err := digest.Validate(); err == nil {
-					images = append(images, "@"+digest.Hex())
-				} else {
-					return errors.Wrapf(err, "error getting config info")
+					return nil, err
 				}
+				pullStructs = append(pullStructs, pullInfo)
+			} else {
+				return nil, errors.Wrapf(err, "error getting config info")
+			}
+		} else {
+			pullInfo, err := r.getPullStruct(srcRef, manifest[0].RepoTags[0])
+			if err != nil {
+				return nil, err
 			}
+			pullStructs = append(pullStructs, pullInfo)
 		}
+
 	} else if srcRef.Transport().Name() == OCIArchive {
 		// retrieve the manifest from index.json to access the image name
 		manifest, err := ociarchive.LoadManifestDescriptor(srcRef)
 		if err != nil {
-			return errors.Wrapf(err, "error loading manifest for %q", srcRef)
+			return nil, errors.Wrapf(err, "error loading manifest for %q", srcRef)
 		}
 
 		if manifest.Annotations == nil || manifest.Annotations["org.opencontainers.image.ref.name"] == "" {
-			return errors.Errorf("error, archive doesn't have a name annotation. Cannot store image with no name")
+			return nil, errors.Errorf("error, archive doesn't have a name annotation. Cannot store image with no name")
+		}
+		pullInfo, err := r.getPullStruct(srcRef, manifest.Annotations["org.opencontainers.image.ref.name"])
+		if err != nil {
+			return nil, err
+		}
+		pullStructs = append(pullStructs, pullInfo)
+	} else if srcRef.Transport().Name() == DirTransport {
+		// supports pull from a directory
+		image := splitArr[1]
+		// remove leading "/"
+		if image[:1] == "/" {
+			image = image[1:]
+		}
+		pullInfo, err := r.getPullStruct(srcRef, image)
+		if err != nil {
+			return nil, err
 		}
-		images = append(images, manifest.Annotations["org.opencontainers.image.ref.name"])
+		pullStructs = append(pullStructs, pullInfo)
 	} else {
-		images = append(images, imgName)
+		pullInfo, err := r.getPullStruct(srcRef, imgName)
+		if err != nil {
+			return nil, err
+		}
+		pullStructs = append(pullStructs, pullInfo)
```

Comment:
(FWIW most of these `append`s could be something like `return []*pullStruct{pullInfo}, nil`; the initial empty value of `pullStruct` is unnecessary.)